src/plugins/process/process.c

```diff
@@ -0,0 +1,286 @@
+	int ret = elektraInvoke1Arg (process->plugin, "open", errorKey);
+	if (ret == -2) return ELEKTRA_PLUGIN_STATUS_SUCCESS;
+	return ret;
+}
+
+int elektraProcessClose (Plugin * handle, Key * errorKey)
+{
+	ElektraPluginProcess * pp = elektraPluginGetData (handle);
+	if (!pp) return ELEKTRA_PLUGIN_STATUS_SUCCESS;
+	Process * process = elektraPluginProcessGetData (pp);
+	if (elektraPluginProcessIsParent (pp))
+	{
+		ElektraPluginProcessCloseResult result = elektraPluginProcessClose (pp, errorKey);
+		if (result.cleanedUp)
+		{
+			cleanup (process, errorKey);
+			elektraPluginSetData (handle, NULL);
+		}
+		return result.result;
+	}
+
+
+	// Handle may be null if initialization failed in the child, ignore that it will exit anyway afterwards
+	if (process->plugin == NULL) return ELEKTRA_PLUGIN_STATUS_SUCCESS;
+
+	int ret = elektraInvoke1Arg (process->plugin, "close", errorKey);
+	if (ret == -2) return ELEKTRA_PLUGIN_STATUS_SUCCESS;
+	return ELEKTRA_PLUGIN_STATUS_SUCCESS;
+}
+
+static void adjustContract (KeySet * pluginContract, KeySet * contract)
+{
+	ksRewind (pluginContract);
+	Key * cur;
+	while ((cur = ksNext (pluginContract)) != NULL)
+	{
+		Key * cpy = keyDup (cur);
+		keySetBaseName (cpy, NULL);
+		if (!elektraStrCmp ("infos", keyBaseName (cpy)))
+		{
+			keySetBaseName (cpy, NULL);
+			keySetBaseName (cpy, NULL);
+			keyAddBaseName (cpy, "process");
+			keyAddBaseName (cpy, "infos");
+			keyAddBaseName (cpy, keyBaseName (cur));
+			Key * infoKey = ksLookup (contract, cpy, KDB_O_NONE);
+			keySetString (infoKey, keyString (cpy));
+		}
+		keyDel (cpy);
+	}
+}
+
+
+int elektraProcessGet (Plugin * handle, KeySet * returned, Key * parentKey)
+{
+	if (isContractKey (parentKey))
+	{
+		KeySet * processConfig = elektraPluginGetConfig (handle);
+		Key * pluginName = ksLookupByName (processConfig, "/plugin", KDB_O_NONE);
+
+		KeySet * contract =
+			ksNew (30, keyNew ("system/elektra/modules/process", KEY_VALUE, "process plugin waits for your orders", KEY_END),
+			       keyNew ("system/elektra/modules/process/exports", KEY_END),
+			       keyNew ("system/elektra/modules/process/exports/open", KEY_FUNC, elektraProcessOpen, KEY_END),
+			       keyNew ("system/elektra/modules/process/exports/close", KEY_FUNC, elektraProcessClose, KEY_END),
+			       keyNew ("system/elektra/modules/process/exports/get", KEY_FUNC, elektraProcessGet, KEY_END),
+			       keyNew ("system/elektra/modules/process/exports/set", KEY_FUNC, elektraProcessSet, KEY_END),
+			       keyNew ("system/elektra/modules/process/exports/error", KEY_FUNC, elektraProcessError, KEY_END),
+			       keyNew ("system/elektra/modules/process/exports/checkconf", KEY_FUNC, elektraProcessCheckConfig, KEY_END),
+#include ELEKTRA_README (process)
+			       keyNew ("system/elektra/modules/process/infos/version", KEY_VALUE, PLUGINVERSION, KEY_END), KS_END);
+		ksAppend (returned, contract);
+		ksDel (contract);
+
+		if (!validPluginName (pluginName, parentKey)) return ELEKTRA_PLUGIN_STATUS_SUCCESS;
```

Comment:
Please comment what is done here?

Shouldn't the contract be renamed to pluginName?

Reply:
as said below, it seems to be expected to be called process